### ISSUE TITLE
Silly party robes available incorrectly

### ISSUE THREAD

**Issue by mxrowanthomas** [HabitRPG/habitica#4601]
I joined habitrpg after the last birthday bash so should only have got the Absurd party robes - instead both were available. I 'bought' both of them mostly as a test which worked at the time, they disappeared over night (possibly a fix) but were available and purchasable again this morning. 

OS x V 10.9.5 Firefox 35.0.1 (up do date) UID: c03e41bd-501f-438c-9553-a7afdf52a08c

**Comment by Alys** [comment]
Thanks @numberland. We're working on the issue. It's taking some time because the database is large and updates are slow to run.

**Comment by Alys** [comment]
@numberland : There's a surprising amount of complexity behind this issue, and we're going to start documenting and discussing everything here. I apologise in advance for the possibly large amount of email updates you might get. There's an unsubscribe button in the right-hand column of this issue, if you get sick of the emails.

**Comment by Alys** [comment]
@lemoness @lefnire  @sabrecat:

I've been running scripts to try to remove the Silly Party Robes from everyone who signed up after the previous birthday celebration (i.e., from people who did not have the Absurd Party Robes), but I've recently realised that those people have been able to buy them back again because of this code: https://github.com/HabitRPG/habitrpg-shared/blob/develop/script/content.coffee#L194
At this point, I'm no longer sure how the robes are meant to work, so I'm backing away from them slowly and trying not to make any sudden movements. :)

What I'm doing instead is making sure that all the `achievements.habitBirthdays` values are correct -- everyone who had an account during 2014 birthday should have a value of 2, everyone else should have a value of 1. Once I'm certain those values are correct, we can use those values to make changes to the robes, if we still want to do that.

We can't use the old `achievements.habitBirthday` boolean flag to correctly adjust `achievements.habitBirthdays` because it's not reliable. I've seen true values for people with account creation dates after the 2014 birthday event. So, I'm deleting that value from all accounts.

Fortunately, we can use the account creation date `auth.timestamps.created` to work out who should have which achievements. So, I'm adjusting the new `achievements.habitBirthdays` numerical value by  setting it to 2 if `auth.timestamps.created` <= 2014-02-01, and setting it to 1 if `auth.timestamps.created` > 2014-02-01.

BUT I'm sick of each of my database scripts taking up to two hours to run, so I'm making them update only those users who have logged in in the past six weeks (`auth.timestamps.loggedin` > 2014-12-20). This means that all our current active users will have correct values for the achievements very soon. I'll fix up the other users later.

**Comment by Alys** [comment]
I'm pretty sure the new `achievements.habitBirthdays` numerical values are now correct for all active users. So, anyone can use that achievement value to control which gear is given to or removed from a user. But I'll double-check the achievements after a nap.

**Comment by SabreCat** [comment]
I'm taking the event duration thing out of the silly party robes now. You're 100% right, they should have been set up like this year's NYE hats, but I added them to the content list more like the first year's.

**Comment by SabreCat** [comment]
My fix is live. Do we have a script ready to run to fix robe ownership once and for all?

**Comment by Alys** [comment]
Script is running to remove items.gear.owned.armor_special_birthday2015 from active users who have a value of 1 for achievements.habitBirthdays

**Comment by Alys** [comment]
The script has finished.

This whole new strategy will fail for some users because the `auth.timestamps.created` field isn't consistent. Some users have a format of `ISODate("2014-01-06T17:54:59.489Z")` while some have `2013-01-10T06:34:12.804Z`. I've only just realised that the latter case exists. My script to set `achievements.habitBirthdays` to either 1 or 2 was based on a date comparison that works for the first format but not for the second. This means that any users that have the second format for their  `auth.timestamps.created` field will have the wrong value for `achievements.habitBirthdays` and hence might not have the correct equipment. 

There's also a third date format `1371974573714` (milliseconds since 1970) but I think that was handled correctly by my script.

I don't know how many users this affects. I suspect it's just some of the really old users and so maybe there's not many of them. Stupidly, I am one of the users it affects, so you'd think I would have noticed this before now. The effect is that `items.gear.owned.armor_special_birthday2015` was removed from my account when it should not have been (because `achievements.habitBirthdays` was never set to 2, because my `auth.timestamps.created` failed the date comparison). I didn't notice this sooner because all my testing was idiotically assuming that the date comparison was working correctly. 

I am sorry but I cannot fix this now. I am exhausted and my head isn't clear enough to think any more. I can't even face uploading the scripts I've been using; there's a few versions now and I don't trust myself to pick out the correct ones, and I can't bring myself to look through them and work out which ones are still usable.

tl;dr Some users - probably just early sign-ups - probably do not have the Silly Party Robes any more. I think the majority of users will be okay. I suppose it depends on how widespread the date format inconsistency is.

@lefnire, please help.

**Comment by paglias** [comment]
@Alys @lefnire if you want to try it later I would suggest to convert dates to a common format using MomentJS, you can pass it the format you have and it'll convert it to a standad Date objects:

`moment(dateInSeveralFormats).toDate()`

I would run it by myself but not having the scripts that have been used and not being familiar with what has been going on I would make thing worse

**Comment by Alys** [comment]
Thanks @paglias. I think I can work on it again today, and that will be useful.

**I have worked out that I can find dates in the mongodb database that are stored as strings rather than as date objects by putting this in the query portion of a find():**
 `'auth.timestamps.created':{$not: {$lt:new Date('2018-01-01')}}`
where 2018-01-01 can be any date in the future. Documenting here in case it's useful in future.

There are 96 users who have no creation date listed at all. Unless anyone tells me otherwise, I am going to assign them a creation date of 2013-01-02T00:00:00.000Z which is a few hours earlier than the earliest value of creation date that we already have.

There are 4924 users who have their creation dates stored in this string format: 2013-02-04T20:37:28.933Z (that's the latest date that uses that format). 

There are 34868 users who have their creation dates stored as milliseconds since 1970 ('1378436245937') - those dates range from 2013-02-05 to 2013-09-08.

I am going to convert those latter two date formats to proper date objects.

@lemoness : As has been mentioned elsewhere, the "In honor of your second day with us..." email was sent to users who had been here for a long time. I believe it probably went out to all 4924 users with the 2013-02-04T20:37:28.933Z string format date. I know it went to me and to Veovis Muad'dib, and we both have that date format. It might also have gone to the 34868 users who have the milliseconds date format - paglias might be able to confirm that. Let me know if you want to know which users they all were; I am saving their UUIDs. Once I've finished fixing these incorrect date formats, emails being incorrectly sent to old users shouldn't happen again.

EDIT: Out of all those users with creation dates in incorrect formats (or missing), only 2619 have logged in in the past six weeks.

**Comment by Alys** [comment]
Update about the 96 users who have no creation date listed at all: At least one of those users is a recent sign-up -- they do not exist in the database backup that I downloaded on 2015-01-11. Hence I will not yet assign creation dates to those accounts. First I want to find out how many of them are recent, and then make a better decision about what date to give them - probably 2013-01-02T00:00:00.000Z for those that aren't in the oldest database backup we have, and the date of the oldest backup for those that are.

**Comment by Alys** [comment]
All users that had string values for `auth.timestamps.created` now have date objects for those same dates.

**Comment by Alys** [comment]
For the 96 users that had no values at all for `auth.timestamps.created`, I've decided not to use the oldest database backup we have, because it takes 2 hours for a backup to download, and another 1.5 hours to import it. Instead, I'm using the copy I already have locally, which was created on 2015-01-11. 

Out of the 96 users that have no creation date recorded, 5 of them do not exist in that database. I.e., in the past three weeks, five new accounts have been created without `auth.timestamps.created` being assigned. **This is disturbing and we should work out why it happens.** I am wondering if it's related to duplicate accounts and will look into that.

The 5 users are 1a0d4b75-73ed-4937-974d-d504d6398884, 1c7ebe27-1250-4f95-ba10-965580adbfd7, 5f972121-4a6d-411c-95e9-7093d3e89b66, ae85818a-e336-4ccd-945e-c15cef975102, ba273976-d9fc-466c-975f-38559d34a824.

I have given them a value of ISODate("2015-01-10T14:00:00Z") for `auth.timestamps.created`.

For the other 91 users that had no values at all for `auth.timestamps.created`, I have given them ISODate("2013-01-02T00:00:00Z"). I still have a record of their UUIDs so I can change that date if required.

EDIT: These two commands (probably) show that there are no more non-date-object or missing created dates:
db.users.find({ 'auth.timestamps.created':{$not: {$lt:new Date('2018-01-01')}} });
db.users.find({ 'auth.timestamps.created':{$exists:false} });

**Comment by Alys** [comment]
All users should now have the correct values for `achievements.habitBirthdays` and all users who were here for more than a year should have the Silly Party Robes. No users who were here for less than a year should now have those Robes. This includes users who had not previously had those things fixed up earlier in the weekend -- i.e., users who had corrupted values for `auth.timestamps.created` (all fixed now) and users who had not logged in for the past six weeks (which I had initially ignored to save time while updating the active users).

In theory, I should double-check all this in a day or two when my head is clearer but I might not bother unless we get reports of oddness.

**Comment by paglias** [comment]
@Alys @lemoness regarding 1 day email, it didn't go out to all those without a date as `auth.timestamps.created` but only to those of them who had cron in the previous days so it's should be a a really small number of them

**Comment by lemoness** [comment]
Should the emails work as expected going forward, or are there still likely
to be glitches?

**Comment by paglias** [comment]
@lemoness i missed this commento, sorry! They're going forward without problems and it should continue to be so
